migrate-to-w3up.js

```diff
@@ -0,0 +1,417 @@
+  if (value instanceof Map) {
+    return Object.fromEntries(value.entries())
+  }
+  if (value instanceof W32023Upload) {
+    return (new W32023UploadSummary(value)).toJSON()
+  }
+  if ((value instanceof Error) && (!('toJSON' in value) || typeof value.toJSON !== 'function')) {
+    return {
+      name: value.name,
+      message: value.message,
+      cause: value.cause,
+    }
+  }
+  return value
+}
+
+/**
+ * get a Space by using interactive cli prompts using inquirer
+ * @param {URL} w3upUrl - url to w3up
+ */
+async function promptForSpace(w3upUrl) {
+  const w3up = await getDefaultW3up(w3upUrl)
+
+  const selection = await select({
+    message: 'choose a space',
+    pageSize: 32,
+    choices: w3up.spaces().map(s => {
+      return {
+        name: [s.name, s.did()].filter(Boolean).join(' - '),
+        value: s,
+        description: JSON.stringify(s.meta)
+      }
+    })
+  })
+  return selection
+}
+
+/**
+ * get a stream of w32023 uploads via
+ * interactive prompts using inquirer
+ * + old web3.storage client library
+ * @returns {Promise<AsyncIterable<W32023Upload> & { length: Promise<number> }>} uploads
+ */
+async function getUploadsFromPrompts() {
+  const confirmation = await confirm({
+    message: 'no uploads were piped in. Do you want to migrate uploads from old.web3.storage?',
+  })
+  if (!confirmation) throw new Error('unable to find a source of uploads to migrate')
+  const envToken = process.env.WEB3_TOKEN
+  let token;
+  if (envToken && await confirm({ message: 'found WEB3_TOKEN in env. Use that?' })) {
+    token = envToken
+  } else {
+    token = await promptForPassword({
+      message: 'enter API token for old.web3.storage',
+    })
+  }
+  const oldW3 = new Web3Storage({ token })
+  const uploads = (async function * () {
+    for await (const u of oldW3.list()) {
+      if (u) yield new W32023Upload(u)
+    }
+  }())
+  // get count
+  const userUploadsResponse = fetch(`https://api.web3.storage/user/uploads`, {
+    headers: { authorization: `Bearer ${token}`},
+  })
+  const count = userUploadsResponse.then(r => parseInt(r.headers.get('count'), 10)).then(c => isNaN(c) ? undefined : c)
+  return Object.assign(uploads, { length: count })
+}
+
+// multicodec codec for CAR bytes
+const CAR_CODE = 0x0202
+
+/**
+ * Attempts to extract a CAR CID from a bucket key.
+ * @param {string} key - string to parse to a CAR CID. e.g. a carpark bucket key basename
+ */
+const stringToCarCid = key => {
+  let errParseCid
+  try {
+    // recent buckets encode CAR CID in filename
+    const cid = Link.parse(key).toV1()
+    return cid
+  } catch (err) {
+    errParseCid = err
+  }
+  // older buckets base32 encode a CAR multihash <base32(car-multihash)>.car
+  // try to parse as base32
+  let errParseBase32
+  try {
+    const digestBytes = fromString(key, 'base32')
+    const digest = Digest.decode(digestBytes)
+    return Link.create(CAR_CODE, digest)
+  } catch (error) {
+    errParseBase32 = error
```

Comment:
What is `errParseCid` and `errParseBase32` doing? They are unused?

Reply:
I changed this to include them in the error.cause that gets thrown if a CID cannot be created from the input

Reply:
https://github.com/web3-storage/migrate-to-w3up/pull/3/commits/0a3418abf910b266331f7b6902e6024f3d90f95e